orttraining/orttraining/core/optimizer/shape_optimizer.cc

```diff
@@ -0,0 +1,355 @@
+  if (!is_scalar) {
+    constant_tensor_proto.add_dims(length);
+  }
+  constant_tensor_proto.set_raw_data(values.data(), length * sizeof(int64_t));
+
+  // Add initializer into Graph.
+  graph.AddInitializedTensor(constant_tensor_proto);
+
+  // Update the output arg shape.
+  ONNX_NAMESPACE::TensorShapeProto new_shape;
+  if (!is_scalar) {
+    new_shape.add_dim()->set_dim_value(length);
+  }
+  arg_to_update->SetShape(new_shape);
+}
+
+}  // namespace
+
+Status ShapeOptimizer::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger)
+    const {
+  GraphViewer graph_viewer(graph);
+  auto& order = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (NodeIndex i : order) {
+    auto* node = graph.GetNode(i);
+    if (!node) {
+      continue;
+    }
+
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(*node, "Shape", {1, 13, 15})) {
+      continue;
+    }
+
+    ORT_RETURN_IF_ERROR(Recurse(*node, modified, graph_level, logger));
+
+    auto data_shape = node->MutableInputDefs()[0]->Shape();
+    if (data_shape == nullptr) {
+      SO_LOG_DEBUG_INFO(logger, "Shape node's data input shape is missing." + node->Name());
+      continue;
+    }
+
+    // Parse data input shape, fill -1 for symbolic dimensions.
+    TensorShapeVector dim_values;
+    dim_values.reserve(data_shape->dim_size());
+    bool has_concrete_dim = false;
+    for (int dim_index = 0; dim_index < data_shape->dim_size(); dim_index++) {
+      auto dim = data_shape->dim(dim_index);
+      if (utils::HasDimValue(dim)) {
+        dim_values.push_back(dim.dim_value());
+        has_concrete_dim = true;
+      } else {
+        // Fill with -1 for symbolic dimension.
+        dim_values.push_back(-1);
+      }
+    }
+
+    if (!has_concrete_dim) {
+      SO_LOG_DEBUG_INFO(logger, "No concrete dimension found, don't need try further." + node->Name());
+      continue;
+    }
+
+    InlinedVector<Node*> nodes_to_remove;
+    TensorShapeVector fold_values;
+    // Short path - check if the shape node can be constant folded.
+    if (CanShapeNodeBeReplacedWithConstant(*node, dim_values, fold_values)) {
+      SO_LOG_DEBUG_INFO(logger, "Shape node can be constant folded." + node->Name());
+      UpdateNodeArgToConstant(graph, node->MutableOutputDefs()[0], fold_values);
+      nodes_to_remove.push_back(node);
+    } else {
+      // Check consumers of Shape node, try best effort to constant fold them if possible.
+      // Currently support Gather and Slice in some cases.
+      auto p_ip_node = node->OutputNodesBegin();
+      const auto p_ip_node_end = node->OutputNodesEnd();
+      InlinedHashSet<const Node*> visited_nodes;
+      while (p_ip_node != p_ip_node_end) {
+        if (visited_nodes.find(&(*p_ip_node)) != visited_nodes.end()) {
+          // Already handled, skip the node.
+          ++p_ip_node;
+          continue;
+        }
+
+        auto& output_node = const_cast<Node&>(*p_ip_node);
```

Comment:
optional: Is const there for a reason? Shall we use `GetMutableConsumerNodes` instead?